### DISCUSSION TITLE
AI & Vega Strike Coding...

### DISCUSSION THREAD

**Discussion by BenjamenMeyer** [vegastrike/Vega-Strike-Engine-Source#887] [Polls]
Generative AI tools such as GitHub Copilot, Google Gemini, and ChatGPT are becoming popular among developers. They can certainly make some coding scenarios easier to get through quickly. However, they also come with risks. Given we have both code and other kinds of assets (documentation, images, etc) we probably should apply the policy evenly.

1. Liability - the licensing for usage of the tools such as GitHub Copilot generally only provides indemnification of copyright violations by the AI to those that paid for the tool.
2. Many open source projects do not allow code submissions (PRs, etc) by AI tooling.
3. We need to have a policy on contributions - this would not apply to AIs built into the system (e.g using AI to live generate text for story lines, auto-build imagery during game play, etc) that the devs decide to build in as it would be an AI controlled by the dev team.

**Comment by royfalk** [comment]
Im thinking of trying co-pilot to refactor existing/generate new content on the assets side. 

I assume this would be ok. 
@BenjamenMeyer ?

**Reply by BenjamenMeyer** [reply]
Copilot is pretty good; its more a liability issue for us. I've thought about it as well; but for the moment let's hold off.
Before we start accepting AI assisted code we probably need to implement commit signing so if there were an issue we could trace it back to a developer for the liability aspect. Thoughts?

We can't 100% avoid it as even if we put something in the PR about it it's on someone's honor to be honest about it.

The other side of it is if we had a licenses issued for the VS org we could allow core developers to do it; but that would require being able to spend a decent chunk each month ($10 or $20 per month depending on the level; though the $10 plan could be free if they considered us popular...which we probably don't qualify for yet).

**Reply by royfalk** [reply]
So, to clarify: 
When **editing** code, it may suggest stuff. One liners and such - I think this is fine. This isn't significantly different from the existing code completion and I doubt anyone could claim this to be a liability. Where I do think care should be given is when writing new code. If co-pilot suggests a block of code, this could be bad.
Opening chat and saying, write some code to draw a sheep - bad.

**Reply by stephengtuggy** [reply]
That makes sense to me. The same would go for JetBrains IDEs' "full line code completion" feature, presumably.

**Reply by BenjamenMeyer** [reply]
@royfalk agreed - for small things it may be hard to tell the difference with liabilities and such. I think I'd like to get to a point where we could accept contributions with AI assistance; it's just a matter of formulating a policy on it. One potential route may be adopting commit level signing...hence the discussion below.

Also, VS Code and some other tools are starting to enable things like GH Copilot by default - which I'm primarily aware of as I see people installing it at work and then asking why they're getting error messages about such functionality and when we tell them to turn it off or get a license from corporate it solves their problems. So it's going to be harder and harder to get contributors without having some kind of policy in place on it.

Liability mostly comes from the training data. Tools like Google Gemini use very limited training data as its more corporate focused and uses only what you tell it; where as Copilot trains on open source software and especially for companies gives a large liability; less so for open source software aside from its lack of distinction about licensing. So GPL software could potentially be pushed into MIT licensed software; not such a big deal the other way; though if having MPL, Apache licensed software pushed to GPL software that could be an issue.

Ultimately I think it's a matter of figuring out how to trace the liability for us - and then hoping we never need to.

**Reply by evertvorster** [reply]
"Hoping we never need to"

In my opinion, AI learns much the same way that people do, by reading material and "remembering" it. Of course there are people who put information on the internet and then expect to be paid every time that someone reads it, and of course they will try to extract money from AI. 
However, AI does not copy information, it reads the information and adjusts some variables and then discards the info. Else the AI models would just be a search engine with a copy of the internet, which it most definitely is not. 

So, this is purely my opinion, but I don't think that the people trying to charge for the information that the AI has used to learn have a strong case. 

AI has one important distinction from humans at this point in time. It has no needs or wants, and it is only us that tell it do think about something and then give an opinion. So, it is the human needs and wants that direct AI. It is our human imagination that directs it. 

As a case in point, I have a project in Github that is entirely AI generated. 
https://github.com/evertvorster/dynamic-power-daemon
It is a tiny utility that monitors the load on my system, and the processes that are running, and dynamically adjusts the power profiles and CPU frequency ranges based on that information. 

If you were to look at the code, would you be able to tell that it was made by an AI? The project only exists because I had a need for it, the AI would not have made it otherwise. So, in effect, I still own that code, because I paid an AI actor to write it, just like a boss at a software company. I picked the license, and it was my idea. 

When you pay a human actor to do the same thing, do you also have to pay for every bit of education that the human has had in order to be able to write that code?

**Reply by BenjamenMeyer** [reply]
@evertvorster I've heard that at least in the US there's a judicial ruling that AI work cannot be copyrighted. Don't know how long that will last; but I do know from work that liability is an issue, and I know from other sources that open source licensed code has been introduced across licenses (e.g to non-open source work) via AI assistance tools like Copilot.

> If you were to look at the code, would you be able to tell that it was made by an AI? The project only exists because I had a need for it, the AI would not have made it otherwise. So, in effect, I still own that code, because I paid an AI actor to write it, just like a boss at a software company. I picked the license, and it was my idea.

The issue is of a quagmire than that. Courts use comparative analysis techniques to compare code when someone is claiming copyright infringement. It is not that it is spotted in the while, but that some says "hey, that's mine and you shouldn't be using it", and you don't know its going to happen until it happens.

Some projects, like WineHQ, go to extreme lengths to protect themselves with respect to intellectual property. Others don't. Given some of our history we probably are somewhere in the middle - projects like PWCU probably need to be more like WineHQ while the general engine can probably be freer. Still, it is a thought on _my_ mind at least, and I'd like to ensure that project is around for others for a long time.

**Comment by stephengtuggy** [comment]
We're already using Dependabot for automated dependency updates. Dependabot is arguably an AI tool. Although I don't think it's quite the same thing as something like Copilot. Just worth considering.

**Reply by BenjamenMeyer** [reply]
Dependabot is mostly reporting security issues, and updating dependencies. It's not writing code like Copilot, Gemini, etc are.

**Reply by stephengtuggy** [reply]
Well, not to the same extent, anyway.

**Comment by BenjamenMeyer** [comment]
Code-Signing: Should we adopt requiring commits to be signed?

**Reply by stephengtuggy** [reply]
I.e. GPG or SSH signatures? I'd be OK with that.

**Reply by stephengtuggy** [reply]
I already do sign most of my commits anyway.

**Reply by royfalk** [reply]
I don't have an opinion. I assume this is a one time effort to setup. What's the advantage of signing here?

**Reply by stephengtuggy** [reply]
It provides a higher degree of certainty that the commits that claim to have been made by you were actually created with your knowledge and consent. Yes, it's a one-time effort to set up.

**Reply by BenjamenMeyer** [reply]
@stephengtuggy PGP signing

@royfalk it's one-time setup similar to SSH keys - you may want/need to rotate them on a periodic basis.
Well, I wrote something up then looked at the `git commit --help` and the `--signoff` option which I think puts it better than what I wrote:
```
       -s, --signoff, --no-signoff
           Add a Signed-off-by trailer by the committer at the end of the commit log message. The meaning of a signoff depends on the project to which you’re committing. For example, it
           may certify that the committer has the rights to submit the work under the project’s license or agrees to some contributor representation, such as a Developer Certificate of
           Origin. (See https://developercertificate.org for the one used by the Linux kernel and Git projects.) Consult the documentation or leadership of the project to which you’re
           contributing to understand how the signoffs are used in that project.

           The --no-signoff option can be used to countermand an earlier --signoff option on the command line.
```
Reading https://developercertificate.org/ I would be in favor of adopting that as part of this, especially if we were to use it as a means of allowing AI contributions. This would essentially be how we could validate the contributor had an appropriate license for their AI usage.

https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

Thoughts?

**Reply by royfalk** [reply]
~I'm fine with signing code.~ _See below_

I'm not sure how you could push a commit to a project without auth. 
I just checked and I have a passkey, which IIRC seems very similar to this. 

After reading the following [s/o question](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for) I finally get it. This is about affirming that I stand behind the commit. **I vote nay on this**. 

I wouldn't trust someone's commit to not contain AI code or infringing code simply because he signed it with GPG. If we want to include AI and are worried about the consequences, we need a real process in place. Policy, procedures, training, approval, etc. Only approved contributors would be allowed this.

**Reply by stephengtuggy** [reply]
@royfalk PGP or GPG signing is an additional step beyond simply committing or pushing a commit. It verifies that the commit is unchanged from what the author intended. It causes the commit to show up in GitHub with a green box next to it with the word `Verified`.

Sign-off is still another, additional step besides committing, pushing, and PGP signing. Sign-off can be said to be about affirming that I stand behind the commit. The sign-off process can be automated, although many sources will say that that defeats the purpose. PGP signing, on the other hand, can and should be automated.

Ultimately, we cannot 100.0000000000% trust any commit not to be nefarious. But these mechanisms can measurably increase the level of certainty, from, say, 96% to 99.999% or something like that. I vote that we adopt at least the PGP signing process for that reason. It will be a one-time effort to set up (unless you're a glutton for punishment, and have it prompt you for the PGP password every time, like me 😉  ) and it will definitely increase the level of trust that we and others can place in our code. Why wouldn't you want that?

**Reply by stephengtuggy** [reply]
Please note, also, that once you have PGP signing set up once on your local machine, it will apply to all your commits across all your repositories, not just the Vega-Strike-related ones. Depending on how you configure it, of course. This could have additional benefits for these other repos, as well.

**Reply by BenjamenMeyer** [reply]
PGP signing will apply at the commit level, and yes shows up in GitHub with the green check. It's enabled through the `signingKey` option in `.gitconfig`  and the `gpgSign` options on commit and tag. I may have gotten the parameters wrong in my earlier message - these are what I setup to enable it.

Sign-off is usually used when a multi-tiered structure is used where we accept commits outside the primary tree and then  sign-off to accept them and push them into the main tree. This is how Linux sends stuff up to Linus. Each maintainers signs off on their section which then goes into a changeset pushed that eventually makes its way to Linus. We're accepting patches through other means, but not the way Linux does... we're not structured that way so I don't think the `sign-off` is useful to us right now.

For the purposes of this conversation I'm more interested in being able to say `change X certainly came from person Y`, at least if they want to use AI Assistance. Thoughts?

**Reply by stephengtuggy** [reply]
Yes, I vote that we adopt PGP signing, but not sign-off. PGP signing can give us something like 99.999% certainty that `change X came from person Y`, irrespective of whether person Y chose to use AI tools for assistance. I think that is about the best we can do for now. Along with some rudimentary, common-sense visual checks to see if the code shows obvious signs of being AI-generated.

**Comment by stephengtuggy** [comment]
What should the end result of this discussion be? Shall we compile a list of approved and non-approved AI's that we can accept contributions from, and under what circumstances and following which guidelines?

**Reply by BenjamenMeyer** [reply]
From my perspective, having a plan to make AI assistance accessible would be acceptable.

**Reply by royfalk** [reply]
So I've been playing around with co-pilot for the last month or two. My conclusions:
1. Line/block suggestions are hit or miss. A lot of times it's scarily intuitive and guesses where I'm going with very little input. The rest of the time it's way of course. 
2. I've had mixed success with stand alone python programs. I asked it to take C++ code and give me the reverse in python and it did so really well. Added the original code and test code. But, where the online documentation is wrong, it will write non-working code.
3. I don't think any of the AI have multi-file awareness. This means that asking it to add code to the engine is a non-starter.

Summary:
1. At least for engine, I think any code that can actually build is fine. I would add some disclaimer regarding IP and  sign the commit to certify it's OK. However, nothing stops people from adding proprietary code manually as well.
2. If someone does add something to the engine through AI **or just negligence**, I don't think legally it would be a big problem. This is because we have no assets or organization. We would be required to remove the offending code. 
3. In any case, AI pulls code from open source. The typical risk is poisoning proprietary code with it not the reverse.

With the rest of the world moving toward this, I'm no longer sure we can avoid it. How long until S/O answers will be modified by AI? Google AI has already answered some of my searches with (wrong and broken) code samples.

BTW, I am not endorsing this or happy with the state of affairs. It is what it is. I don't think we can craft a policy to completely mitigate the risk.

**Reply by BenjamenMeyer** [reply]
> I don't think any of the AI have multi-file awareness. This means that asking it to add code to the engine is a non-starter.

I'm aware that best practice is to have a lot of files open as it will pull from any files you have open in addition to what it was trained on. So there is some multi-file awareness but it won't just randomly add files, even if it sees include/import statements for them.

> BTW, I am not endorsing this or happy with the state of affairs. It is what it is. I don't think we can craft a policy to completely mitigate the risk.

Agreed, which is why my goal is to have some kind of policy on it.

**Reply by BenjamenMeyer** [reply]
> If someone does add something to the engine through AI or just negligence, I don't think legally it would be a big problem. This is because we have no assets or organization. We would be required to remove the offending code.

The risk for us is getting delisted from GitHub, mirroring to BitBucket/Gitlab might help but even that won't be full-proof.

**Comment by stephengtuggy** [comment]
Started a new, follow-up poll here: https://github.com/orgs/vegastrike/discussions/1491

**Comment by BenjamenMeyer** [comment]
Adding something to the CONTRIBUTING guidelines in https://github.com/vegastrike/Vega-Strike-Engine-Source/pull/1669
Let's creating something practical there. once that gets merged I'll copy it over to the other repositories.
